### PR TITLE
feat: PFCP heartbeats and headless mode

### PR DESCRIPTION
**Pull request by marknefedov**

### CODE REVIEW

File: cmd/core/node_association.go [hunk 1]
```diff
@@ -1,15 +1,62 @@
...
+	"time"
+)
 
 type NodeAssociation struct {
 	ID            string
 	Addr          string
 	NextSessionID uint64
-	Sessions      SessionMap
+	Sessions      map[uint64]*Session
+	HbRetries     uint32
```
Comment (pirog-spb) [inline cmd/core/node_association.go:14]: А есть ли смысл сокращать? Почему не `HeartbeatRetries`?

File: cmd/core/pfcp_connection.go [hunk 2]
```diff
@@ -57,6 +64,12 @@ func CreatePfcpConnection(addr string, pfcpHandlerMap PfcpHandlerMap, nodeId str
 }
 
 func (connection *PfcpConnection) Run() {
+	go func() {
+		for {
+			connection.ProcessHeadlessAssociations()
+			time.Sleep(time.Duration(config.Conf.HeartBeatInterval) * time.Second)
```
Comment (pirog-spb) [inline cmd/core/pfcp_connection.go:70]: `HeartBeatInterval` -> HeartbeatInterval

File: cmd/core/pfcp_connection.go [hunk 3]
```diff
@@ -101,3 +114,42 @@ func (connection *PfcpConnection) SendMessage(msg message.Message, addr *net.UDP
...
+		}
+	}
+}
+
+func (connection *PfcpConnection) DeleteAssociation(assocAddr string) {
+	assoc := connection.GetAssociation(assocAddr)
+	log.Printf("Pruning expired node association: %s", assocAddr)
+	for sessionId, session := range assoc.Sessions {
+		log.Printf("Deleting session: %d", sessionId)
+		for _, far := range session.FARs {
```
Comment (pirog-spb) [inline cmd/core/pfcp_connection.go:135]: Может выделить всё что в цикле в метод `connection.DeleteSession()`?

File: cmd/core/pfcp_hearbeat.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+func HandlePfcpHeartbeatResponse(conn *PfcpConnection, msg message.Message, addr string) (message.Message, error) {
+	hbresp := msg.(*message.HeartbeatResponse)
+	ts, err := hbresp.RecoveryTimeStamp.RecoveryTimeStamp()
+	if err != nil {
+		log.Printf("Got Heartbeat Response with invalid TS: %s, from: %s", err, addr)
+	} else {
+		log.Printf("Got Heartbeat Response with TS: %s, from: %s", ts, addr)
+	}
+	if association := conn.GetAssociation(addr); association != nil {
```
Comment (pirog-spb) [inline cmd/core/pfcp_hearbeat.go:37]: Why in response handler GetAssociation is called after recovery timestamp processing?

File: cmd/core/pfcp_hearbeat.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+func SendHearbeatReqeust(conn *PfcpConnection, addr string) {
+	hbreq := message.NewHeartbeatRequest(0, ie.NewRecoveryTimeStamp(time.Now()), nil)
+	log.Printf("Sent Heartbeat Request to: %s", addr)
+	udpAddr, err := net.ResolveUDPAddr("udp", addr+":8805")
+	if err == nil {
+		if err := conn.SendMessage(hbreq, udpAddr); err != nil {
+			log.Printf("Failed to send heartbeat request: %s\n", err.Error())
+		}
+	} else {
+		log.Printf("Failed to send Hearbit Request: %s\n", err.Error())
```
Comment (pirog-spb) [inline cmd/core/pfcp_hearbeat.go:52]: Typo

File: cmd/core/node_association.go [hunk 6]
```diff
@@ -1,15 +1,62 @@
...
+		Sessions:      make(map[uint64]*Session),
+	}
 }
 
 func (association *NodeAssociation) NewLocalSEID() uint64 {
 	association.NextSessionID += 1
 	return association.NextSessionID
 }
+
+func (association *NodeAssociation) CheckInContact() {
```
Comment (pirog-spb) [inline cmd/core/node_association.go:32]: `CheckInContact` -> `Refresh`/`Update`?

File: cmd/core/pfcp_connection.go [hunk 7]
```diff
@@ -101,3 +114,42 @@ func (connection *PfcpConnection) SendMessage(msg message.Message, addr *net.UDP
...
 	return nil
 }
+
+func (connection *PfcpConnection) ProcessHeadlessAssociations() {
+	for assocAddr, assoc := range connection.NodeAssociations {
+		if assoc.IsExpired() {
+			log.Printf("Pruning expired node association: %s", assocAddr)
+			connection.DeleteAssociation(assocAddr)
+		}
+		if assoc.cancelRetries == nil {
```
Comment (pirog-spb) [inline cmd/core/pfcp_connection.go:124]: Looks strange. Seems like should be called only if association is not expired. I'd proposed extract method like `assoc.IsHeartbeatScheduled`

File: cmd/core/pfcp_connection.go [hunk 8]
```diff
@@ -101,3 +114,42 @@ func (connection *PfcpConnection) SendMessage(msg message.Message, addr *net.UDP
...
+			SendTimeoutHeartbeatRequests(time.Duration(config.Conf.HeartBeatTimeout)*time.Second, connection, assocAddr)
```
Comment (pirog-spb) [inline cmd/core/pfcp_connection.go:125]: `SendTimeoutHeartbeatRequests` -> `ScheduleHeartbeatRequest`

File: cmd/core/pfcp_connection.go [hunk 9]
```diff
@@ -101,3 +114,42 @@ func (connection *PfcpConnection) SendMessage(msg message.Message, addr *net.UDP
 	}
 	return nil
 }
+
+func (connection *PfcpConnection) ProcessHeadlessAssociations() {
```
Comment (pirog-spb) [inline cmd/core/pfcp_connection.go:118]: Actually process every association. So I propose to name it like `UpdateAssociations`/`RefreshAssociations`

File: cmd/core/pfcp_hearbeat.go (hunk 4 above)
Comment (marknefedov) [reply cmd/core/pfcp_hearbeat.go:38]: I've forgotten a return if timestamp is corrupted.

File: cmd/core/pfcp_connection.go [hunk 10]
```diff
@@ -10,18 +11,24 @@ import (
...
+	NodeAssociations map[string]*NodeAssociation
 	nodeId           string
 	nodeAddrV4       net.IP
 	n3Address        net.IP
 	mapOperations    ebpf.ForwardingPlaneController
 }
 
+func (connection *PfcpConnection) GetAssociation(assocAddr string) *NodeAssociation {
+	assoc, ok := connection.NodeAssociations[assocAddr]
+	if ok {
```
Comment (pirog-spb) [inline cmd/core/pfcp_connection.go:26]: Let's try conbine 25 and 26 lines in one if statement?

File: cmd/core/node_association.go [hunk 11]
```diff
@@ -1,15 +1,67 @@
...
+func (association *NodeAssociation) IsExpired() bool {
+	return association.HeartbeatRetries > config.Conf.HeartbeatRetries
+}
+
+func (association *NodeAssociation) IsHeartbeatScheduled() bool {
+	return association.cancelRetries != nil
+}
+
+// ScheduleHeartbeatRequest schedules a series of heartbeat requests to be sent to the remote node. Return a cancellation function to stop the scheduled requests.
+func ScheduleHeartbeatRequest(duration time.Duration, conn *PfcpConnection, association string) context.CancelFunc {
```
Comment (pirog-spb) [inline cmd/core/node_association.go:49]: Seems like `ScheduleHeartbeatRequest` and `SendHeartbeatRequest` belong to accosiation type and have to be methods of this type

File: cmd/core/pfcp_connection.go [hunk 12]
```diff
@@ -101,3 +114,48 @@ func (connection *PfcpConnection) SendMessage(msg message.Message, addr *net.UDP
...
 }
+
+// RefreshAssociations checks for expired associations and schedules heartbeats for those that are not expired.
+func (connection *PfcpConnection) RefreshAssociations() {
+	for assocAddr, assoc := range connection.NodeAssociations {
+		if assoc.IsExpired() {
+			log.Printf("Pruning expired node association: %s", assocAddr)
+			connection.DeleteAssociation(assocAddr)
+		}
+		if !assoc.IsHeartbeatScheduled() {
```
Comment (pirog-spb) [inline cmd/core/pfcp_connection.go:125]: What if association has already expired and deleted in prev lines?

File: cmd/core/pfcp_connection.go [hunk 13]
```diff
@@ -101,3 +113,50 @@ func (connection *PfcpConnection) SendMessage(msg message.Message, addr *net.UDP
 	}
 	return nil
 }
+
+// RefreshAssociations checks for expired associations and schedules heartbeats for those that are not expired.
+func (connection *PfcpConnection) RefreshAssociations() {
+	for assocAddr, assoc := range connection.NodeAssociations {
+		if assoc.IsExpired() {
+			log.Printf("Pruning expired node association: %s", assocAddr)
+			connection.DeleteAssociation(assocAddr)
```
Comment (pirog-spb) [inline cmd/core/pfcp_connection.go:122]: Are you sure that it's safe to delete association just in the middle of for cicle?

Comment (marknefedov) [reply cmd/core/pfcp_connection.go:122]: This is a map and not an array/list so, it should be fine.
https://groups.google.com/g/golang-nuts/c/eJqwQLONhLs

> If map entries that have not yet been reached are deleted during iteration, the
corresponding iteration values will not be produced.

**Review by pirog-spb — APPROVED** [review]

**Pull request merged**